jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java

```diff
@@ -719,7 +699,9 @@ public boolean checkVirtualHost(Request request)
         if (_vhosts.isEmpty())
             return true;
 
-        String host = normalizeHostname(request.getHttpURI().getHost());
+        // TODO check why normalizeVirtualHostname is necessary to pass ContextHandlerCollectionTest
+        //      specifically is it valid for a request hostname to end with '.'
+        String host = normalizeVirtualHostname(Request.getServerName(request));
```

Comment:
@joakime can you recall why we allow trailing dots in request hostnames?

Reply:
Turns out a dot at the end of a host name tells the DNS resolver to not use search domains.

So for "http://foo/path", "foo" needs to be resolved and if your `/etc/resolv.conf` has `search domain.com`, then `foo.domain.com` is looked up.

But for `http://foo./path`, no search domains are used, so that is already a fully qualified domain name.

Reply:
The trailing dot indicates a fully qualified name (host name / dns / resolver / windows machine names / netbios names / etc).

It prevents the lookup of the name within a resolver search list. (eg: the `search` configuration in `/etc/resolv.conf`)

That trailing dot also refers to the root of the dns hierarchy.

This means when you use `https://webtide.com/` the lookup on your linux machine is ...

1. the `/etc/hosts` file
2. the `/etc/resolv.conf` search list (eg: `webtide.com.localdomain` is looked up)
3. the DNS hierarchy
  a. the `com` TLD
  b. the `webtide` domain

But if you use `https://webtide.com./` then the search is only ever `webtide.com` and that name is never considered a sub-domain of any searched domains.

Reply:
The DNS protocol uses the trailing dot as well.

``` plain
$ dig jetty.org
; <<>> DiG 9.18.18-0ubuntu0.22.04.1-Ubuntu <<>> jetty.org
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 7287
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;jetty.org.			IN	A

;; ANSWER SECTION:
jetty.org.		233	IN	A	107.22.210.238

;; Query time: 0 msec
;; SERVER: 127.0.0.53#53(127.0.0.53) (UDP)
;; WHEN: Tue Nov 28 12:11:07 CST 2023
;; MSG SIZE  rcvd: 54
```

Notice the trailing dot in the dns record?

Even the MX records show this trailing dot.

``` plain
$ dig jetty.org MX

; <<>> DiG 9.18.18-0ubuntu0.22.04.1-Ubuntu <<>> jetty.org MX
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 21841
;; flags: qr rd ra; QUERY: 1, ANSWER: 5, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;jetty.org.			IN	MX

;; ANSWER SECTION:
jetty.org.		215	IN	MX	10 alt3.aspmx.l.google.com.
jetty.org.		215	IN	MX	5 alt2.aspmx.l.google.com.
jetty.org.		215	IN	MX	1 aspmx.l.google.com.
jetty.org.		215	IN	MX	5 alt1.aspmx.l.google.com.
jetty.org.		215	IN	MX	10 alt4.aspmx.l.google.com.

;; Query time: 4 msec
;; SERVER: 127.0.0.53#53(127.0.0.53) (UDP)
;; WHEN: Tue Nov 28 12:11:25 CST 2023
;; MSG SIZE  rcvd: 156

```